Code/GraphMol/RGroupDecomposition/RGroupDecomp.cpp

```diff
@@ -444,58 +444,89 @@ std::vector<std::string> RGroupDecomposition::getRGroupLabels() const {
           usedLabelMap.has(label) && usedLabelMap.isUserDefined(label);
-      auto numExplicitHs = nbrAtom->getNumExplicitHs();
-      if (usedLabelMap.has(label) && usedLabelMap.getIsUsed(label)) {
-        if (numExplicitHs) {
-          nbrAtom->setNumExplicitHs(numExplicitHs - 1);
-        }
-      } else if (!isUserDefinedLabel ||
-                 data->params.removeAllHydrogenRGroupsAndLabels) {
-        coreWithMatches->removeAtom(atomIdx);
-        atomRemoved = true;
-        // if we remove an unused label from an aromatic atom,
-        // we need to check whether we need to adjust its explicit
-        // H count, or it will fail to kekulize
-        if (isUserDefinedLabel && nbrAtom->getIsAromatic() && !missingRGroup) {
-          nbrAtom->updatePropertyCache(false);
-          if (!numExplicitHs) {
-            nbrAtom->setNumExplicitHs(nbrAtom->getExplicitValence() -
-                                      nbrAtom->getDegree());
-          }
-        }
+      const bool isUsedLabel =
+          usedLabelMap.has(label) && usedLabelMap.getIsUsed(label);
+      if (!isUsedLabel && (!isUserDefinedLabel ||
+                           data->params.removeAllHydrogenRGroupsAndLabels)) {
+        // Always convert to hydrogen - then remove later if
+        // removeHydrogensPostMatch is set
+        atom->setAtomicNum(1);
+        atom->updatePropertyCache(false);
+      } else {
+        retainedRGroups[atom] = label;
       }
-      nbrAtom->updatePropertyCache(false);
-      if (!atomRemoved && coreWithMatches->getNumConformers() > 0) {
-        auto iter = match.rgroups.find(label);
-        if (iter != match.rgroups.end() && iter->second->is_hydrogen) {
-          MolOps::setTerminalAtomCoords(*coreWithMatches, atomIdx,
-                                        nbrAtom->getIdx());
-        }
+    }
+  }
+
+#ifdef VERBOSE
+  std::cerr << "output core mol2 " << MolToSmiles(*coreWithMatches)
+            << std::endl;
+#endif
+  if (data->params.removeHydrogensPostMatch) {
+    RDLog::LogStateSetter blocker;
+    constexpr bool implicitOnly = false;
+    constexpr bool updateExplicitCount = false;
+    constexpr bool sanitize = false;
+    MolOps::removeHs(*coreWithMatches, implicitOnly, updateExplicitCount,
+                     sanitize);
+    coreWithMatches->updatePropertyCache(false);
+  }
+
+  if (coreWithMatches->getNumConformers() > 0) {
+    for (const auto & [atom, label] : retainedRGroups) {
+      if (usedLabelMap.has(label) && usedLabelMap.isUserDefined(label)) {
+        // coordinates of user defined R groups should already be copied over
+        continue;
+      }
+      const auto neighbor = *coreWithMatches->atomNeighbors(atom).begin();
+      const auto &mapping = data->finalRlabelMapping;
+      const auto oldLabel =
+          std::find_if(mapping.begin(), mapping.end(),
+                       [label = label](const auto &p) { return p.second == label; });
+      if (auto iter = match.rgroups.find(oldLabel->first);
```

Comment:
don't you need to check whether or not the `find_if` succeeded before you use `oldLabel` here?

Reply:
> don't you need to check whether or not the `find_if` succeeded before you use `oldLabel` here?

I've made that check.